src/internal_telemetry/allocations/mod.rs

```diff
@@ -22,14 +22,25 @@ pub(crate) use self::allocator::{
 
 use crossbeam_utils::CachePadded;
 
-const NUM_GROUPS: usize = 256;
+const NUM_GROUPS: usize = 128;
 /// These arrays represent the allocations and deallocations for each group.
 /// We pad each Atomic to reduce false sharing effects.
-static GROUP_MEM_ALLOCS: [CachePadded<AtomicU64>; NUM_GROUPS] =
-    arr![CachePadded::new(AtomicU64::new(0)); 256];
-static GROUP_MEM_DEALLOCS: [CachePadded<AtomicU64>; NUM_GROUPS] =
-    arr![CachePadded::new(AtomicU64::new(0)); 256];
-
+static GROUP_MEM_STATS: [[CachePadded<AtomicI64>; NUM_GROUPS]; 8] = [
+    arr![CachePadded::new(AtomicI64::new(0)); 128],
+    arr![CachePadded::new(AtomicI64::new(0)); 128],
+    arr![CachePadded::new(AtomicI64::new(0)); 128],
+    arr![CachePadded::new(AtomicI64::new(0)); 128],
+    arr![CachePadded::new(AtomicI64::new(0)); 128],
+    arr![CachePadded::new(AtomicI64::new(0)); 128],
+    arr![CachePadded::new(AtomicI64::new(0)); 128],
+    arr![CachePadded::new(AtomicI64::new(0)); 128],
+];
```

Comment:
I think there's value to having allocations and deallocations be separate, primarily in that it let's us tracking if allocations/deallocations have been happening, even if the two numbers are identical.

For example, without tracking them separately, if we observe the stats after allocating and deallocating 100 bytes, all we know is that our current outstanding allocations has a value of 0... but we can't see that we actually allocated 100 bytes.

It also strikes me that in terms of reducing contention, fully splitting the allocation/deallocation path would be advantageous.

Reply:
One problem w/ separating allocs/deallocs is potential overflow. We can run into a scenario were we do frequent small allocs/deallocs.

Reply:
The reporting thread could also just track the last seen value for a given group IDs allocations/deallocations, and do the requisite math to calculate the delta, while accounting for overflow.

It's also weird that this is now `i64`. If that change was to also more gracefully handle underflow (deallocs > allocs) then we should actually be checking for underflow in the reporter and emitting an error when we detect that... because obviously that's an invariant violation for an allocation group to have negative allocated bytes.

Reply:
It's for a scenario where we allocate in thread X but deallocate in thread Y. In this case, thread Y's value will be negative.

Reply:
Yeah, that's fair.

I still think it's worth experimenting with separating allocations and deallocations to see if it provides a measurable improvement when stacked on top of the sharded/padded statics approach.

Reply:
Should we do that in a follow up PR?

Reply:
We can do that in a follow-up PR, sure.